### PR TITLE
Add normalize step for co/colorado_gov (#262)

### LINKED ISSUE #262: normalize co/colorado_gov
[learn our pipeline: normalize](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Runner-pipeline-stages#normalize)

Blocked by #261

Normalize existing [`.ndjson`](http://ndjson.org/) records into [the Vaccinate The States data schema](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Normalized-Location-Schema).

Read all files in the directory passed as the second argument (`sys.argv[2]`), transform each record into the normalized schema (filling in as much detail as possible), then output them to new files in the directory passed as the first argument (`sys.argv[1]`).

Check the wiki to learn more about the purpose of the normalize stage and how to get set up for development!

### Tips

1. Fetch and parse data for this site before you start developing:
    ```sh
    poetry run vaccine-feed-ingest fetch co/colorado_gov
    poetry run vaccine-feed-ingest parse co/colorado_gov
    ```

1. While working on your code, run it at any point:
    ```sh
    poetry run vaccine-feed-ingest normalize co/colorado_gov
    ```

### Example
[Parsed files for `ak/arcgis`](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/ak/arcgis/parsed) are converted to [normalized files](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/ak/arcgis/normalized)

### PR DESCRIPTION
**Pull request by skbdpup**
# Normalize colorado_gov for CO

| Key Details |
|-|
| Resolves #262 |
State: CO |
Site: colorado_gov |

## Notes
1. For the six Community Vaccination Sites, there is hardly any structured data. Instead of hardcoding contact info, or trying to parse it, I just put the free-form data in the Notes section.
2. There is no contact info for "Federal Retail Pharmacies". I try falling back on a store page on the chain's website, or the generic store locator page.
3. Still, over a dozen locations have no contact info at all. ~I exclude them from the output, but log a warning.~

## Data sample

```json
{"id": "co_colorado_gov:colorado_state_fairgrounds_community_vaccination_site", "name": "Colorado State Fairgrounds: Community Vaccination Site", "address": {"street1": null, "street2": null, "city": null, "state": "CO", "zip": null}, "location": null, "contact": [{"contact_type": "general", "phone": null, "website": null, "email": null, "other": "Hours of Operation: 8 a.m. - 8 p.m.\n\nDays of Operation: Monday - Thursday\n\nRegistration Link:\nwww.Centura.org/vaccine\n\nSame Day and Walk/Drive Up Appointments Available\n\nDedicated call number for any mass vaccine events. Below is the call center number 720-263-5737 that supports questions/registration if guests do not have intranet. This is currently open now from 7:00 a.m. \u2013 8:00 p.m. MST 7 days a week\n\nType of Vaccine: Pfizer"}], "languages": null, "opening_dates": null, "opening_hours": null, "availability": null, "inventory": null, "access": null, "parent_organization": null, "links": null, "notes": null, "active": null, "source": {"source": "co_colorado_gov", "id": "colorado_state_fairgrounds_community_vaccination_site", "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6", "fetched_at": "2021-05-03T04:09:47.566618", "published_at": null, "data": {"_folder_name": "Community Vaccination Site ", "name": "Colorado State Fairgrounds: Community Vaccination Site", "description": ["Hours of Operation: 8 a.m. - 8 p.m.", "", "Days of Operation: Monday - Thursday", "", "Registration Link:", "www.Centura.org/vaccine", "", "Same Day and Walk/Drive Up Appointments Available", "", "Dedicated call number for any mass vaccine events. Below is the call center number 720-263-5737 that supports questions/registration if guests do not have intranet. This is currently open now from 7:00 a.m. \u2013 8:00 p.m. MST 7 days a week", "", "Type of Vaccine: Pfizer"], "styleUrl": "#icon-1502-4E342E", "Point": {"long": "-104.6428985", "lat": "38.2485466"}}}}
{"id": "co_colorado_gov:costco_pharmacy_1030", "name": "Costco Pharmacy # 1030", "address": {"street1": "5050 North Nevada Ave", "street2": null, "city": "Colorado Springs", "state": "CO", "zip": null}, "location": null, "contact": [{"contact_type": "general", "phone": null, "website": "https://www.costco.com/warehouse-locations", "email": null, "other": null}], "languages": null, "opening_dates": null, "opening_hours": null, "availability": null, "inventory": [{"vaccine": "moderna", "supply_level": null}], "access": null, "parent_organization": {"id": "costco", "name": null}, "links": [{"authority": "costco", "id": "1030", "uri": null}], "notes": null, "active": null, "source": {"source": "co_colorado_gov", "id": "costco_pharmacy_1030", "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6", "fetched_at": "2021-05-03T04:09:47.566618", "published_at": null, "data": {"_folder_name": "Federal Retail Pharmacies", "name": "Costco Pharmacy # 1030", "address": "5050 North Nevada Ave, Colorado Springs, CO", "description": ["Mapped Description: Moderna Vaccine", "Address: 5050 North Nevada Ave, Colorado Springs, CO"], "styleUrl": "#icon-1899-A52714", "ExtendedData": {"Mapped Description": "Moderna Vaccine", "Address": "5050 North Nevada Ave, Colorado Springs, CO"}}}}
{"id": "co_colorado_gov:king_soopers_pharmacy_62000073", "name": "King Soopers Pharmacy 62000073", "address": {"street1": "4503 John F Kennedy Pkwy", "street2": null, "city": "Fort Collins", "state": "CO", "zip": null}, "location": null, "contact": [{"contact_type": "general", "phone": null, "website": "https://www.kingsoopers.com/stores/details/620/62000073", "email": null, "other": null}], "languages": null, "opening_dates": null, "opening_hours": null, "availability": null, "inventory": [{"vaccine": "moderna", "supply_level": null}], "access": null, "parent_organization": {"id": "king_soopers", "name": null}, "links": [{"authority": "king_soopers", "id": "62000073", "uri": null}], "notes": null, "active": null, "source": {"source": "co_colorado_gov", "id": "king_soopers_pharmacy_62000073", "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6", "fetched_at": "2021-05-03T04:09:47.566618", "published_at": null, "data": {"_folder_name": "Federal Retail Pharmacies", "name": "King Soopers Pharmacy 62000073", "address": "4503 John F Kennedy Pkwy, Fort Collins, CO", "description": ["Mapped Description: Moderna Vaccine", "Address: 4503 John F Kennedy Pkwy, Fort Collins, CO"], "styleUrl": "#icon-1899-A52714", "ExtendedData": {"Mapped Description": "Moderna Vaccine", "Address": "4503 John F Kennedy Pkwy, Fort Collins, CO"}}}}
{"id": "co_colorado_gov:walmart_pharmacy_3176", "name": "Walmart Pharmacy #3176", "address": {"street1": "1725 N Union Blvd", "street2": null, "city": "Colorado Springs", "state": "CO", "zip": null}, "location": null, "contact": [{"contact_type": "general", "phone": null, "website": "https://www.walmart.com/store/3176", "email": null, "other": null}], "languages": null, "opening_dates": null, "opening_hours": null, "availability": null, "inventory": [{"vaccine": "moderna", "supply_level": null}], "access": null, "parent_organization": {"id": "walmart", "name": null}, "links": [{"authority": "walmart", "id": "3176", "uri": null}], "notes": null, "active": null, "source": {"source": "co_colorado_gov", "id": "walmart_pharmacy_3176", "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6", "fetched_at": "2021-05-03T04:09:47.566618", "published_at": null, "data": {"_folder_name": "Federal Retail Pharmacies", "name": "Walmart Pharmacy #3176", "address": "1725 N Union Blvd, Colorado Springs, CO", "description": ["Mapped Description: Moderna Vaccine", "Address: 1725 N Union Blvd, Colorado Springs, CO"], "styleUrl": "#icon-1899-A52714", "ExtendedData": {"Mapped Description": "Moderna Vaccine", "Address": "1725 N Union Blvd, Colorado Springs, CO"}}}}
{"id": "co_colorado_gov:7033", "name": "Good Day Pharmacy #14", "address": {"street1": "1900 16th St Unit 102", "street2": null, "city": "Greeley", "state": "CO", "zip": null}, "location": null, "contact": [{"contact_type": "general", "phone": "(970) 576-3178", "website": null, "email": null, "other": null}, {"contact_type": "general", "phone": null, "website": "https://www.gooddaypharmacy.com/", "email": null, "other": null}], "languages": null, "opening_dates": null, "opening_hours": null, "availability": null, "inventory": [{"vaccine": "pfizer_biontech", "supply_level": null}], "access": null, "parent_organization": null, "links": null, "notes": null, "active": null, "source": {"source": "co_colorado_gov", "id": "7033", "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6", "fetched_at": "2021-05-03T04:09:47.566618", "published_at": null, "data": {"_folder_name": "Pfizer", "name": "Good Day Pharmacy #14", "address": "1900 16th St Unit 102, Greeley, CO", "description": ["PIN: 7033", "county: Weld", "Description: COV-19 (Pfizer); MDV6; 195-pack", "type: Pharmacy", "address: 1900 16th St Unit 102, Greeley, CO", "phone: (970) 576-3178", "website: https://www.gooddaypharmacy.com/", "vaccine sign up:", "spanishresources:", "registrtiondescription:"], "styleUrl": "#icon-1899-9C27B0", "ExtendedData": {"PIN": "7033", "county": "Weld", "Description": "COV-19 (Pfizer); MDV6; 195-pack", "type": "Pharmacy", "address": "1900 16th St Unit 102, Greeley, CO", "phone": "(970) 576-3178", "website": "https://www.gooddaypharmacy.com/", "vaccine sign up": null, "spanishresources": null, "registrtiondescription": null}}}}
```

<details><summary>Pretty-printed</summary>

```json
{
  "id": "co_colorado_gov:colorado_state_fairgrounds_community_vaccination_site",
  "name": "Colorado State Fairgrounds: Community Vaccination Site",
  "address": {
    "street1": null,
    "street2": null,
    "city": null,
    "state": "CO",
    "zip": null
  },
  "location": null,
  "contact": [
    {
      "contact_type": "general",
      "phone": null,
      "website": null,
      "email": null,
      "other": "Hours of Operation: 8 a.m. - 8 p.m.\n\nDays of Operation: Monday - Thursday\n\nRegistration Link:\nwww.Centura.org/vaccine\n\nSame Day and Walk/Drive Up Appointments Available\n\nDedicated call number for any mass vaccine events. Below is the call center number 720-263-5737 that supports questions/registration if guests do not have intranet. This is currently open now from 7:00 a.m. – 8:00 p.m. MST 7 days a week\n\nType of Vaccine: Pfizer"
    }
  ],
  "languages": null,
  "opening_dates": null,
  "opening_hours": null,
  "availability": null,
  "inventory": null,
  "access": null,
  "parent_organization": null,
  "links": null,
  "notes": null,
  "active": null,
  "source": {
    "source": "co_colorado_gov",
    "id": "colorado_state_fairgrounds_community_vaccination_site",
    "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6",
    "fetched_at": "2021-05-03T04:09:47.566618",
    "published_at": null,
    "data": {
      "_folder_name": "Community Vaccination Site ",
      "name": "Colorado State Fairgrounds: Community Vaccination Site",
      "description": [
        "Hours of Operation: 8 a.m. - 8 p.m.",
        "",
        "Days of Operation: Monday - Thursday",
        "",
        "Registration Link:",
        "www.Centura.org/vaccine",
        "",
        "Same Day and Walk/Drive Up Appointments Available",
        "",
        "Dedicated call number for any mass vaccine events. Below is the call center number 720-263-5737 that supports questions/registration if guests do not have intranet. This is currently open now from 7:00 a.m. – 8:00 p.m. MST 7 days a week",
        "",
        "Type of Vaccine: Pfizer"
      ],
      "styleUrl": "#icon-1502-4E342E",
      "Point": {
        "long": "-104.6428985",
        "lat": "38.2485466"
      }
    }
  }
}
{
  "id": "co_colorado_gov:costco_pharmacy_1030",
  "name": "Costco Pharmacy # 1030",
  "address": {
    "street1": "5050 North Nevada Ave",
    "street2": null,
    "city": "Colorado Springs",
    "state": "CO",
    "zip": null
  },
  "location": null,
  "contact": [
    {
      "contact_type": "general",
      "phone": null,
      "website": "https://www.costco.com/warehouse-locations",
      "email": null,
      "other": null
    }
  ],
  "languages": null,
  "opening_dates": null,
  "opening_hours": null,
  "availability": null,
  "inventory": [
    {
      "vaccine": "moderna",
      "supply_level": null
    }
  ],
  "access": null,
  "parent_organization": {
    "id": "costco",
    "name": null
  },
  "links": [
    {
      "authority": "costco",
      "id": "1030",
      "uri": null
    }
  ],
  "notes": null,
  "active": null,
  "source": {
    "source": "co_colorado_gov",
    "id": "costco_pharmacy_1030",
    "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6",
    "fetched_at": "2021-05-03T04:09:47.566618",
    "published_at": null,
    "data": {
      "_folder_name": "Federal Retail Pharmacies",
      "name": "Costco Pharmacy # 1030",
      "address": "5050 North Nevada Ave, Colorado Springs, CO",
      "description": [
        "Mapped Description: Moderna Vaccine",
        "Address: 5050 North Nevada Ave, Colorado Springs, CO"
      ],
      "styleUrl": "#icon-1899-A52714",
      "ExtendedData": {
        "Mapped Description": "Moderna Vaccine",
        "Address": "5050 North Nevada Ave, Colorado Springs, CO"
      }
    }
  }
}
{
  "id": "co_colorado_gov:king_soopers_pharmacy_62000073",
  "name": "King Soopers Pharmacy 62000073",
  "address": {
    "street1": "4503 John F Kennedy Pkwy",
    "street2": null,
    "city": "Fort Collins",
    "state": "CO",
    "zip": null
  },
  "location": null,
  "contact": [
    {
      "contact_type": "general",
      "phone": null,
      "website": "https://www.kingsoopers.com/stores/details/620/62000073",
      "email": null,
      "other": null
    }
  ],
  "languages": null,
  "opening_dates": null,
  "opening_hours": null,
  "availability": null,
  "inventory": [
    {
      "vaccine": "moderna",
      "supply_level": null
    }
  ],
  "access": null,
  "parent_organization": {
    "id": "king_soopers",
    "name": null
  },
  "links": [
    {
      "authority": "king_soopers",
      "id": "62000073",
      "uri": null
    }
  ],
  "notes": null,
  "active": null,
  "source": {
    "source": "co_colorado_gov",
    "id": "king_soopers_pharmacy_62000073",
    "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6",
    "fetched_at": "2021-05-03T04:09:47.566618",
    "published_at": null,
    "data": {
      "_folder_name": "Federal Retail Pharmacies",
      "name": "King Soopers Pharmacy 62000073",
      "address": "4503 John F Kennedy Pkwy, Fort Collins, CO",
      "description": [
        "Mapped Description: Moderna Vaccine",
        "Address: 4503 John F Kennedy Pkwy, Fort Collins, CO"
      ],
      "styleUrl": "#icon-1899-A52714",
      "ExtendedData": {
        "Mapped Description": "Moderna Vaccine",
        "Address": "4503 John F Kennedy Pkwy, Fort Collins, CO"
      }
    }
  }
}
{
  "id": "co_colorado_gov:walmart_pharmacy_3176",
  "name": "Walmart Pharmacy #3176",
  "address": {
    "street1": "1725 N Union Blvd",
    "street2": null,
    "city": "Colorado Springs",
    "state": "CO",
    "zip": null
  },
  "location": null,
  "contact": [
    {
      "contact_type": "general",
      "phone": null,
      "website": "https://www.walmart.com/store/3176",
      "email": null,
      "other": null
    }
  ],
  "languages": null,
  "opening_dates": null,
  "opening_hours": null,
  "availability": null,
  "inventory": [
    {
      "vaccine": "moderna",
      "supply_level": null
    }
  ],
  "access": null,
  "parent_organization": {
    "id": "walmart",
    "name": null
  },
  "links": [
    {
      "authority": "walmart",
      "id": "3176",
      "uri": null
    }
  ],
  "notes": null,
  "active": null,
  "source": {
    "source": "co_colorado_gov",
    "id": "walmart_pharmacy_3176",
    "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6",
    "fetched_at": "2021-05-03T04:09:47.566618",
    "published_at": null,
    "data": {
      "_folder_name": "Federal Retail Pharmacies",
      "name": "Walmart Pharmacy #3176",
      "address": "1725 N Union Blvd, Colorado Springs, CO",
      "description": [
        "Mapped Description: Moderna Vaccine",
        "Address: 1725 N Union Blvd, Colorado Springs, CO"
      ],
      "styleUrl": "#icon-1899-A52714",
      "ExtendedData": {
        "Mapped Description": "Moderna Vaccine",
        "Address": "1725 N Union Blvd, Colorado Springs, CO"
      }
    }
  }
}
{
  "id": "co_colorado_gov:7033",
  "name": "Good Day Pharmacy #14",
  "address": {
    "street1": "1900 16th St Unit 102",
    "street2": null,
    "city": "Greeley",
    "state": "CO",
    "zip": null
  },
  "location": null,
  "contact": [
    {
      "contact_type": "general",
      "phone": "(970) 576-3178",
      "website": null,
      "email": null,
      "other": null
    },
    {
      "contact_type": "general",
      "phone": null,
      "website": "https://www.gooddaypharmacy.com/",
      "email": null,
      "other": null
    }
  ],
  "languages": null,
  "opening_dates": null,
  "opening_hours": null,
  "availability": null,
  "inventory": [
    {
      "vaccine": "pfizer_biontech",
      "supply_level": null
    }
  ],
  "access": null,
  "parent_organization": null,
  "links": null,
  "notes": null,
  "active": null,
  "source": {
    "source": "co_colorado_gov",
    "id": "7033",
    "fetched_from_uri": "https://www.google.com/maps/d/viewer?mid=1x9KT3SJub0igOTnhFtdRYmceZuBXMWvK&ll=38.98747216882165,-105.9556642978642&z=6",
    "fetched_at": "2021-05-03T04:09:47.566618",
    "published_at": null,
    "data": {
      "_folder_name": "Pfizer",
      "name": "Good Day Pharmacy #14",
      "address": "1900 16th St Unit 102, Greeley, CO",
      "description": [
        "PIN: 7033",
        "county: Weld",
        "Description: COV-19 (Pfizer); MDV6; 195-pack",
        "type: Pharmacy",
        "address: 1900 16th St Unit 102, Greeley, CO",
        "phone: (970) 576-3178",
        "website: https://www.gooddaypharmacy.com/",
        "vaccine sign up:",
        "spanishresources:",
        "registrtiondescription:"
      ],
      "styleUrl": "#icon-1899-9C27B0",
      "ExtendedData": {
        "PIN": "7033",
        "county": "Weld",
        "Description": "COV-19 (Pfizer); MDV6; 195-pack",
        "type": "Pharmacy",
        "address": "1900 16th St Unit 102, Greeley, CO",
        "phone": "(970) 576-3178",
        "website": "https://www.gooddaypharmacy.com/",
        "vaccine sign up": null,
        "spanishresources": null,
        "registrtiondescription": null
      }
    }
  }
}
```

</details>

## Before Opening a PR
- [x] I tested this using the CLI (e.g., `poetry run vaccine-feed-ingest <state>/<site>`)
- [x] I ran auto-formatting: `poetry run tox -e lint-fix`

### CODE REVIEW

File: vaccine_feed_ingest/utils/normalize.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def provider_id_from_name(
     m = re.search(r"BROOKSHIRE PHARMACY #\d+ #(\d+)", name, re.I)
     if m:
         return VaccineProvider.BROOKSHIRE, str(int(m.group(1)))
-    m = re.search(r"COSTCO(?: MARKET)? PHARMACY #(\d+)", name, re.I)
+    m = re.search(r"COSTCO(?: MARKET)? PHARMACY #\s*(\d+)", name, re.I)
```
Comment (eliblock) [inline vaccine_feed_ingest/utils/normalize.py:59]: :|

File: vaccine_feed_ingest/runners/co/colorado_gov/normalize.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+                    schema.Contact(
+                        contact_type=schema.ContactType.BOOKING, website=website
+                    )
+                )
+    else:
+        # Community vaccination sites have no "ExtendedData"; the contact info is in free-form notes
+        contacts.append(
+            schema.Contact(
+                contact_type=schema.ContactType.GENERAL,
+                other="\n".join(site["description"]),
```
Comment (eliblock) [inline vaccine_feed_ingest/runners/co/colorado_gov/normalize.py:157]: The seems like a reasonable fallback. Thank you!

File: vaccine_feed_ingest/runners/co/colorado_gov/normalize.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+def _get_normalized_location(
+    site: dict, timestamp: str
+) -> Optional[schema.NormalizedLocation]:
+    source = "co_colorado_gov"
+    id = _get_id(site)
+
+    contacts = _get_contacts(site)
+    if not contacts:
+        logger.warning("No contacts for %s %s", site["name"], site["address"])
+        return None
```
Comment (eliblock) [inline vaccine_feed_ingest/runners/co/colorado_gov/normalize.py:246]: Please include the sites _even if_ they have no `Contacts`. Part of the goal of this pipeline is to merge available data sources which have different portions of information - it's possible that some other source will have contact info for these sites, but might not have the same notes, or similar.

**Review by eliblock — CHANGES_REQUESTED** [review]
This is a really strong normalizer, and it will be good to get this data online (including the vaccine type availability 🤩 ).

Let's include all the reported sites - even if they are missing contact info.

File: vaccine_feed_ingest/runners/co/colorado_gov/normalize.py (hunk 3 above)
Comment (skbdpup) [reply vaccine_feed_ingest/runners/co/colorado_gov/normalize.py]: Whoops, I mistakenly believed that null contacts were forbidden by the schema. Fixed!

**Pull request merged**